catalyst_voices_packages/catalyst_cardano_serialization/lib/src/scripts.dart

```diff
@@ -0,0 +1,497 @@
+import 'dart:typed_data';
+import 'package:catalyst_cardano_serialization/src/types.dart';
+import 'package:cbor/cbor.dart';
+import 'package:convert/convert.dart';
+import 'package:equatable/equatable.dart';
+import 'package:pinenacl/digests.dart';
+
+/// The size of the script hash in bytes.
+const scriptHashSize = 28;
+
+/// Define an enum for the different types of native scripts
+enum NativeScriptType {
+  /// Public key
+  pubkey(0),
+
+  /// All
+  all(1),
+
+  /// Any
+  any(2),
+
+  /// N of K
+  nOfK(3),
+
+  /// Invalid before
+  invalidBefore(4),
+
+  /// Invalid after
+  invalidAfter(5);
+
+  /// The value of the enum
+  final int value;
+  const NativeScriptType(this.value);
+}
+
+/// Define an enum for the different types of reference scripts
+enum RefScriptType {
+  /// Native script
+  native(0),
+
+  /// PlutusV1 script
+  plutusV1(1),
+
+  /// PlutusV2 script
+  plutusV2(2),
+
+  /// PlutusV3 script
+  plutusV3(3);
+
+  /// The value of the enum
+  final int value;
+  const RefScriptType(this.value);
+}
+
+/// Define a sealed class for scripts that extends Equatable.
+sealed class Script extends Equatable implements CborEncodable {
+  const Script();
+
+  /// Computes the hash of the script.
+  ///
+  /// This method converts the script to its CBOR representation,
+  /// adds a tag based on the script type, and then computes the
+  /// Blake2b hash of the resulting bytes.
+  Uint8List get hash {
+    final tag = switch (this) {
+      NativeScript _ => 0,
+      PlutusV1Script _ => 1,
+      PlutusV2Script _ => 2,
+      PlutusV3Script _ => 3,
+      _ => throw ArgumentError.value(
+          this,
+          'script',
+          'Invalid script type to hash.',
+        ),
+    };
+
+    final cborValue = toCbor();
+    final bytesToHash = cborValue is CborBytes
+        ? _handleDoubleDecodedCbor(cborValue.bytes)
+        : cborValue;
+
+    final cborBytes = cbor.encode(bytesToHash);
+    final bytes = Uint8List.fromList([tag, ...cborBytes]);
+    return Hash.blake2b(bytes, digestSize: scriptHashSize);
+  }
+
+  CborValue _handleDoubleDecodedCbor(List<int> bytes) {
```

Comment:
```suggestion
  CborValue _handleDoubleEncodedCbor(List<int> bytes) {
```

You're trying to decode something that is double encoded, not decoded (if I understood correctly the code)

Reply:
Yeah, you're right.